src/kvp/key.rs

```diff
@@ -0,0 +1,360 @@
+use std::{fmt::Display, ops::Deref, str::FromStr};
+
+use lazy_static::lazy_static;
+use regex::Regex;
+use serde::{de::Visitor, Deserialize, Serialize};
+use snafu::{ensure, ResultExt, Snafu};
+
+const KEY_PREFIX_MAX_LEN: usize = 253;
+const KEY_NAME_MAX_LEN: usize = 63;
+
+lazy_static! {
+    static ref KEY_PREFIX_REGEX: Regex =
+        Regex::new(r"^[a-zA-Z](\.?[a-zA-Z0-9-])*\.[a-zA-Z]{2,}\.?$").unwrap();
+    static ref KEY_NAME_REGEX: Regex =
+        Regex::new(r"^[a-z0-9A-Z]([a-z0-9A-Z-_.]*[a-z0-9A-Z]+)?$").unwrap();
+}
+
+#[derive(Debug, PartialEq, Snafu)]
+pub enum KeyError {
+    #[snafu(display("key input cannot be empty"))]
+    EmptyInput,
+
+    #[snafu(display("invalid number of slashes in key - expected 0 or 1, got {count}"))]
+    InvalidSlashCharCount { count: usize },
+
+    #[snafu(display("failed to parse key prefix"))]
+    KeyPrefixError { source: KeyPrefixError },
+
+    #[snafu(display("failed to parse key name"))]
+    KeyNameError { source: KeyNameError },
+}
+
+/// The [`Key`] of a [`KeyValuePair`](crate::kvp::KeyValuePair). It contains an
+/// optional prefix, and a required name. The Kubernetes documentation defines
+/// the format and allowed characters [here][k8s-labels]. A [`Key`] is always
+/// validated. It also doesn't provide any associated functions which enable
+/// unvalidated manipulation of the inner values.
+///
+/// [k8s-labels]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Key {
+    prefix: Option<KeyPrefix>,
+    name: KeyName,
+}
+
+impl FromStr for Key {
+    type Err = KeyError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let input = input.trim();
+
+        // The input cannot be empty
+        ensure!(!input.is_empty(), EmptyInputSnafu);
+
+        // Split the input up into the optional prefix and name
+        let parts: Vec<_> = input.split('/').collect();
+
+        // Ensure we have 2 or less parts. More parts are a result of too many
+        // slashes
+        ensure!(
+            parts.len() <= 2,
+            InvalidSlashCharCountSnafu {
+                count: parts.len() - 1
+            }
+        );
+
+        let (prefix, name) = if parts.len() == 1 {
+            (None, KeyName::from_str(parts[0]).context(KeyNameSnafu)?)
+        } else {
+            (
+                Some(KeyPrefix::from_str(parts[0]).context(KeyPrefixSnafu)?),
+                KeyName::from_str(parts[1]).context(KeyNameSnafu)?,
+            )
+        };
+
+        Ok(Self { prefix, name })
```

Comment:
```suggestion
let (prefix, name) = if let Some((prefix, name)) = input.split_once('/') {
    ensure!(!name.contains('/'), NestedPrefixSnafu);
    (Some(prefix), name)
} else {
    (None, input)
}

Ok(Self {
    prefix: prefix.map(KeyPrefix::from_str).transpose().context(KeyPrefixSnafu)?,
    name: KeyName::from_str(name).context(KeyNameSnafu)?,
})
```

Reply:
Or alternatively,

```rust
let parts = input.split('/').collect::<Vec<_>>();
match &parts[..] {
  [name] => (None, name),
  [prefix, name] = (Some(prefix), name),
  _ => return NestedPrefixSnafu.fail(),
}
```

Reply:
But this splitting also seems like a case of "why are we parsing these consts instead of just constructing valid data to begin with"?

Reply:
Changed in b3a897e